Add save clips to Blink services

perdue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The blinkpy library has been updated to support locally stored video clips.
Blink's local storage mechanism is very slow.  They choose to upload
a clip to their cloud from local storage, and then download it back down.
As a result of this, race conditions make it impossible to get at all the clips
a camera captures.  To mitigate this, new functionality has been added to allow
a user to download all recently-captured videos and download them on a schedule.

Library changes:
https://github.com/fronzbot/blinkpy/commit/47faf5a0ab848805b75e34fd6c996fdfb9f21c9a

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #82454
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25329

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

perdue commented:
I am a first-time contributor.  Who is a maintainer that can approve running the workflows?

perdue commented:
@Petro31 in the [forums](https://community.home-assistant.io/t/wth-why-does-it-take-more-than-half-a-year-to-get-your-pr-reviewed-by-a-core-developer/471647/3) says to comment in the PR to keep it fresh and towards the top.

I would love to get this PR reviewed soon.

frenck commented:
> @Petro31 in the [forums](https://community.home-assistant.io/t/wth-why-does-it-take-more-than-half-a-year-to-get-your-pr-reviewed-by-a-core-developer/471647) says to comment in the PR to keep it fresh and towards the top.

Please don't, it doesn't work like that. Additionally, please only merge the dev branch, if needed. I didn't see any need in all previous merges.

Additionally, consider using rebasing instead of merging (as per our developer documentation: <https://developers.home-assistant.io/docs/development_catching_up>)

../Frenck

frenck commented on homeassistant/components/blink/manifest.json:
```diff
@@ -2,7 +2,7 @@
   "name": "Blink",
   "documentation": "https://www.home-assistant.io/integrations/blink",
-  "requirements": ["blinkpy==0.19.2"],
+  "requirements": ["blinkpy==0.20.0.rc0"],
```
Why is it RC? Unless we are fixing a completely broken integration and have no other way, we do not allow the use of dev/alpha/beta/release candidates.

perdue replied:
Understood.

The original Blink integration is not broken.  But the `0.19.x` versions do not work _at all_ for those of use who don't use Blink's cloud storage.  In that sense, it is broken/unsupported.

@fronzbot :  Are you ready to release `blinkpy` updates as `v0.20.0`?

perdue replied:
This PR is now using the full release version `0.20.0`.  It is ready for another review @frenck.  Thanks!

perdue commented:
> Please don't, it doesn't work like that. Additionally, please only merge the dev branch, if needed. I didn't see any need in all previous merges.

OK.  Please forgive my ignorance here.  I am used to projects that require you to stay up-to-date with the target branch.

> Additionally, consider using rebasing instead of merging (as per our developer documentation: https://developers.home-assistant.io/docs/development_catching_up)

Gotcha.  I usually rebase after addressing any PR comments.  I have simply been clicking "Update branch" on the GitHub UI.

perdue commented:
@frenck The PR branch is rebased onto `dev` now.  It looks like the workflow must be approved again.

perdue committed:
Add save clips to Blink services

d4icon reviewed (approved):
I have the custom integration of these changes and so far they are working correctly.
I give it a thumbs up!

cacordina commented:
> custom integration of these changes and so far they are work

Same on my side, I use it since 2 weeks and it works correctly

perdue commented:
@frenck Would you have some time to review this PR?  There are now two other community testers running it as a custom integration with success.  Thanks!

perdue commented:
I added another Blink camera to my system.  As a result, I was automatically signed up for another 30-day cloud storage trial from Blink.  I can confirm that the integration/library works with cloud storage and local storage.  I've been running this code as a custom integration for around 4 months now.

@frenck Could you or someone review this PR?  I would like to get this merged in.  Thanks!

eoaddai commented:
Can't wait for this to be available on production. its the solution i've been waiting for.

raman325 commented:
all, please be patient, and do not tag people to ask for reviews. As you can see, there is a long backlog, and the team moves as quickly as we can

raman325 commented on homeassistant/components/blink/alarm_control_panel.py:
```diff
@@ -55,8 +55,15 @@ def __init__(self, data, name, sync):
         """Update the state of the device."""
-        _LOGGER.debug("Updating Blink Alarm Control Panel %s", self._name)
-        self.data.refresh()
+        if self.data.check_if_ok_to_update():
```
what does this do?

perdue replied:
The blink integration was calling refresh multiple times.  This uses a `blinkpy` method to check if an update should be called based on the last refresh time.

https://github.com/fronzbot/blinkpy/blob/a40011cb74e6e110cd8a5330e3879e4ac7a50c79/blinkpy/blinkpy.py#L295

perdue commented on homeassistant/components/blink/__init__.py:
```diff
@@ -164,13 +184,33 @@ async def async_handle_save_video_service(hass, entry, call):
+        _LOGGER.error("Can't write to directory %s, no access to path!", clips_dir)
+        return
+
+    def _save_recent_clips(name, output_dir):
```
That is the crux of the PR.  The `_write_video` only triggers a single video (the most recent) to be written.  I left it in so that current users' automations do not break.

The new "save recent clips" adds a service for saving all clips generated since the last refresh.

perdue replied:
As far as your question about access to `hass`.  I followed the same pattern as the `async_handle_save_video_service` function.  I'm not trying to do a refactor here.

raman325 replied:
thanks for clarifying

perdue commented:
> all, please be patient, and do not tag people to ask for reviews. As you can see, there is a long backlog, and the team moves as quickly as we can

@raman325 Thanks for the review!  I have addressed your comments.

raman325 commented:
can be merged once tests pass

perdue commented:
> Hello, I was trying to run this under Developer Tools > Services after I whitelisted my directory (/config/www/blink) but no files are saved. I was wondering if this is live?

Hi @brylee123. The original functionality for the Blink integration is live.  This means that if you have the cloud storage option, you should be able to save the most recent clip to your directory.  If you have Blink local storage, then we are still waiting for this PR to be merged and released.

perdue commented:
The PR was approved 3 weeks ago, so hopefully the codebase maintainers will merge it soon.

perdue commented:
> I added another Blink camera to my system. As a result, I was automatically signed up for another 30-day cloud storage trial from Blink. I can confirm that the integration/library works with cloud storage and local storage. I've been running this code as a custom integration for around 4 months now.
> 

My 30-day cloud storage trial expired, and the updated integration (running as a custom one until this PR gets merged) switched to using the local storage (_i.e._, Sync Module usb drive) seamlessly.  :partying_face:

frenck commented:
@perdue The PR was approved 2 weeks ago, but there is a merge conflict, so it cannot be merged. Could you resolve it? Thanks! 👍

frenck commented:
I've marked this PR draft, as changes are needed. Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

perdue committed:
Merge branch 'dev' into perdue/feature/add_save_clips_service_to_blink

perdue commented:
> I've marked this PR draft, as changes are needed. Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.
> 
> Thanks! +1
> 
> ../Frenck
> 
> [_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

@frenck Conflicts have been resolved.  Ready for review.  Please enable the workflow again.  Thanks!

perdue commented:
Note:  By merging [this other PR](https://github.com/home-assistant/core/pull/93735), you have also merged in all of my updates to `blinkpy` that support the local storage functionality.  So thanks for that!  Those changes were included in version `0.20.0`.

Now this PR is reduced to adding in the HA service call to "save recent clips", which allows a user to retrieve more than just a single recent clip.

frenck reviewed (approved):
LGTM, thanks, @perdue 👍 

../Frenck
